.github/workflows/imperative.yml

```diff
@@ -0,0 +1,75 @@
+# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
+
+name: Imperative CI
+
+on: [push]
+
+jobs:
+  test:
+
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        node-version: [8.x, 10.x, 12.x]
+        os: [windows-latest, ubuntu-latest, macos-latest]
+
+    env:
+      OS: ${{ matrix.os }}
+      NODE: ${{ matrix.node-version }}
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+
+    - name: Use Node.js ${{ matrix.node-version }}
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+
+    - name: Install Node Package Dependencies
+      id: install
+      run: npm ci
+
+    - name: Build Source
+      id: build
+      run: npm run build --if-present
+
+    - name: Unit Tests
+      id: unit
+      if: ${{ always() && steps.build.outcome == 'success' }}
+      run: npm run test:unit >> file.txt
+
+    - name: Unlock Keyring
+      if: ${{ always() && steps.build.outcome == 'success' && matrix.os == 'ubuntu-latest' }}
+      uses: tjohnsonBCM/unlock-keyring@master
+
+    - name: Integration Tests
+      id: integration
+      if: ${{ always() && steps.build.outcome == 'success' }}
+      run: npm run test:integration >> file.txt
+
+    - name: Zip Results
+      id: zip
+      if: ${{ always() && steps.build.outcome == 'success' }}
+      uses: papeloto/action-zip@v1
+      with:
+        files: __tests__/__results__/
+        dest: results.zip
+        recursive: false
+
+    - name: Archive Results
+      id: upload
+      if: ${{ always() && steps.build.outcome == 'success' }}
```

Comment:
For my own education, what happens without the check for `always() && steps.build.outcome == 'success' `

Reply:
It depends.  Without `always()`, it won't run when the tests fail, which is exactly when we want to look at the artifacts for the tests. However, then it will try to archive artifacts if the build failed, and they won't be present, so we added the extra logic to make sure that it built.

Reply:
thanks - this is good knowledge for the zowe explorer workflow!